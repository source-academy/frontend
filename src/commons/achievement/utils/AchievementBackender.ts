import {
  AchievementGoal,
  AchievementItem,
  GoalDefinition,
  GoalMeta,
  GoalType
} from '../../../features/achievement/AchievementTypes';

export const backendifyGoalDefinition = (goal: GoalDefinition) => ({
  targetCount: goal.meta.type === GoalType.ASSESSMENT ? 1 : goal.meta.targetCount,
  meta: goal.meta,
  text: goal.text,
  type: goal.meta.type,
  uuid: goal.uuid
});

export const frontendifyAchievementGoal = (goal: any) =>
  ({
    uuid: goal.uuid || '',
    text: goal.text || '',
    achievementUuids: goal.achievementUuids,
    meta: (goal.meta.type === 'Event'
      ? {
          ...goal.meta,
          release: goal.meta.release ? new Date(goal.meta.release) : undefined,
          deadline: goal.meta.deadline ? new Date(goal.meta.deadline) : undefined,
          observeFrom: goal.meta.observeFrom ? new Date(goal.meta.observeFrom) : undefined,
          observeTo: goal.meta.observeTo ? new Date(goal.meta.observeTo) : undefined
        }
      : goal.meta) as GoalMeta,
    count: goal.count,
    targetCount: goal.targetCount,
    completed: goal.count >= goal.targetCount
  } as AchievementGoal);

export const frontendifyAchievementItem = (achievement: any) =>
  ({
    uuid: achievement.uuid || '',
    title: achievement.title || '',
    xp: achievement.xp,
    isVariableXp: achievement.isVariableXp,
    deadline: achievement.deadline === null ? undefined : new Date(achievement.deadline),
    release: achievement.release === null ? undefined : new Date(achievement.release),
    isTask: achievement.isTask,
    position: achievement.position,
    prerequisiteUuids: achievement.prerequisiteUuids,
    goalUuids: achievement.goalUuids,
    cardBackground: achievement.cardBackground || '',
    view: {
      coverImage: achievement.view.coverImage || '',
      completionText: achievement.view.completionText || '',
      description: achievement.view.description || ''
    }
  } as AchievementItem);

export const backendifyAchievementItem = (achievement: AchievementItem) => ({
  ...achievement,
  deadline: achievement.deadline ? achievement.deadline : null,
  release: achievement.release ? achievement.release : null
});
