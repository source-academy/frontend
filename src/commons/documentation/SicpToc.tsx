import { Tree, TreeNodeInfo } from '@blueprintjs/core';
import { cloneDeep } from 'lodash';
import React, { useState } from 'react';
import toc from 'src/features/sicp/data/toc.json';

type Props = {
  handleCloseToc?: () => void;
  handleNodeClicked?: (node: TreeNodeInfo) => void;
};

/**
 * Table of contents of SICP.
 */
const SicpToc: React.FC<Props> = props => {
  const [sidebarContent, setSidebarContent] = useState(toc as TreeNodeInfo[]);

  const handleNodeExpand = (_node: TreeNodeInfo, path: integer[]) => {
    const newState = cloneDeep(sidebarContent);
    Tree.nodeFromPath(path, newState).isExpanded = true;
    setSidebarContent(newState);
  };

  const handleNodeCollapse = (_node: TreeNodeInfo, path: integer[]) => {
    const newState = cloneDeep(sidebarContent);
    Tree.nodeFromPath(path, newState).isExpanded = false;
    setSidebarContent(newState);
  };

  return (
    <div className="sicp-toc sicp-toc-assessment">
      <Tree
        className="sicp-toc-tree"
        contents={sidebarContent}
        onNodeClick={props.handleNodeClicked}
        onNodeCollapse={handleNodeCollapse}
        onNodeExpand={handleNodeExpand}
      />
    </div>
  );
};

export default SicpToc;
