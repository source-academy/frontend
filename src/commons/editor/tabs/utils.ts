import path from 'path';

/**
 * Returns the shortest file paths that is uniquely identifiable among
 * all open editor tabs. This is similar to how most code editors available
 * handle the displaying of file names.
 *
 * For example, if there are 3 open editor tabs where the file names are
 * exactly the same, we would need to display more of the file path to be
 * able to distinguish which editor tab corresponds to which file. Given
 * the following absolute file paths:
 * - /a.js
 * - /dir1/a.js
 * - /dir1/dir2/a.js
 * The shortest unique file paths will be:
 * - /a.js
 * - /dir1/a.js
 * - /dir2/a.js
 *
 * @param originalFilePaths The file paths to be shortened.
 */
export const getShortestUniqueFilePaths = (originalFilePaths: string[]): string[] => {
  // Store the unique shortened file paths as a mapping from the original file paths
  // to the shortened file paths. This is necessary because the output of this function
  // must preserve the original ordering of file paths.
  const originalToUniqueShortenedFilePaths: Record<string, string> = {};
  // Split each original file path into path segments and store the mapping from file
  // path to path segments for O(1) lookup. Since we only deal with the BrowserFS file
  // system, the path separator will always be '/'.
  const filePathSegments: Record<string, string[]> = originalFilePaths.reduce(
    (segments, filePath) => ({
      ...segments,
      // It is necessary to remove empty segments to deal with the very first '/' in
      // file paths.
      [filePath]: filePath.split(path.sep).filter(segment => segment !== '')
    }),
    {}
  );

  for (
    let numOfPathSegments = 1;
    // Keep looping while some original file paths have yet to be shortened.
    Object.keys(originalToUniqueShortenedFilePaths).length < originalFilePaths.length;
    numOfPathSegments++
  ) {
    // Based on the number of path segments for the iteration, we construct the
    // shortened file path. We then store the mapping from the shortened file path
    // to any original file path which transforms into it.
    const shortenedToOriginalFilePaths: Record<string, string[]> = Object.entries(
      filePathSegments
    ).reduce((filePaths, [originalFilePath, filePathSegments]) => {
      // Note that if there are fewer path segments than the number being sliced,
      // all of the path segments will be returned without error.
      const shortenedFilePath = '/' + filePathSegments.slice(-numOfPathSegments).join('/');
      return {
        ...filePaths,
        [shortenedFilePath]: (filePaths[shortenedFilePath] ?? []).concat(originalFilePath)
      };
    }, {});
    // Each shortened file path that only has a single corresponding original file
    // path is added to the unique shortened file paths record and their entry in
    // the file path segments record is removed to prevent further processing.
    Object.entries(shortenedToOriginalFilePaths).forEach(
      ([shortenedFilePath, originalFilePaths]) => {
        if (originalFilePaths.length > 1) {
          return;
        }

        const originalFilePath = originalFilePaths[0];
        originalToUniqueShortenedFilePaths[originalFilePath] = shortenedFilePath;
        // Remove the file path's segments from the next iteration.
        delete filePathSegments[originalFilePath];
      }
    );
  }

  // Finally, we retrieve the unique shortened file paths while preserving the ordering
  // of file paths.
  return originalFilePaths.map(filePath => originalToUniqueShortenedFilePaths[filePath]);
};
