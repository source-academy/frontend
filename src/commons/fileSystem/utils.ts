import { FSModule } from 'browserfs/dist/node/core/FS';
import Stats from 'browserfs/dist/node/core/node_fs_stats';
import path from 'path';

import { WORKSPACE_BASE_PATHS } from '../../pages/fileSystem/createInBrowserFileSystem';
import { WorkspaceLocation } from '../workspace/WorkspaceTypes';

type File = {
  path: string;
  contents: string;
};

/**
 * Retrieves the files in the specified workspace as a record that maps from
 * file path to file content. Because BrowserFS lacks an equivalent to Node.js's
 * Promises API, we need to wrap each asynchronous call to the file system with
 * promises ourselves.
 *
 * @param workspaceLocation The location of the workspace.
 * @param fileSystem        The file system instance.
 */
export const retrieveFilesInWorkspaceAsRecord = (
  workspaceLocation: WorkspaceLocation,
  fileSystem: FSModule
): Promise<Record<string, string>> => {
  const processFile = (filePath: string): Promise<File> => {
    return new Promise((resolve, reject) => {
      fileSystem.readFile(filePath, 'utf-8', (err, fileContents) => {
        if (err) {
          reject(err);
          return;
        }
        if (fileContents === undefined) {
          return;
        }

        resolve({
          path: filePath,
          contents: fileContents
        });
      });
    });
  };

  const processDirectory = (directoryPath: string): Promise<File[]> => {
    return new Promise((resolve, reject) => {
      fileSystem.readdir(directoryPath, (err, fileNames) => {
        if (err) {
          reject(err);
          return;
        }
        if (fileNames === undefined) {
          return;
        }

        const promises: Array<Promise<File | File[]>> = [];
        for (const fileName of fileNames) {
          const fullPath = path.join(directoryPath, fileName);
          promises.push(
            new Promise((resolve, reject) => {
              fileSystem.lstat(fullPath, (err, stats) => {
                if (err) {
                  reject(err);
                  return;
                }
                if (stats === undefined) {
                  return;
                }

                if (stats.isFile()) {
                  resolve(processFile(fullPath));
                } else if (stats.isDirectory()) {
                  resolve(processDirectory(fullPath));
                }
              });
            })
          );
        }

        Promise.all(promises).then((values: Array<File | File[]>) => {
          resolve(values.flat());
        });
      });
    });
  };

  const files = processDirectory(WORKSPACE_BASE_PATHS[workspaceLocation]);
  // Convert from array to record.
  return files.then((files: File[]) => {
    return files.reduce((record: Record<string, string>, file: File) => {
      record[file.path] = file.contents;
      return record;
    }, {});
  });
};

/**
 * Overwrites the files in the workspace with the ones specified. Because BrowserFS
 * lacks an equivalent to Node.js's Promises API, we need to wrap each asynchronous
 * call to the file system with promises ourselves.
 *
 * @param workspaceLocation The location of the workspace.
 * @param fileSystem        The file system instance.
 * @param files             A mapping from file paths to file contents.
 */
export const overwriteFilesInWorkspace = (
  workspaceLocation: WorkspaceLocation,
  fileSystem: FSModule,
  files: Record<string, string>
): Promise<void> => {
  return rmFilesInDirRecursively(fileSystem, WORKSPACE_BASE_PATHS[workspaceLocation]).then(() => {
    return new Promise(async (resolve, reject) => {
      for (const [filePath, fileContents] of Object.entries(files)) {
        await writeFileRecursively(fileSystem, filePath, fileContents).catch(err => reject(err));
      }
      resolve();
    });
  });
};

/**
 * Removes the files & directories in a directory recursively, but leave
 * the directory itself intact.
 *
 * @param fileSystem    The file system instance.
 * @param directoryPath The path of the directory to be removed.
 */
export const rmFilesInDirRecursively = (
  fileSystem: FSModule,
  directoryPath: string
): Promise<void> => {
  return new Promise((resolve, reject) => {
    fileSystem.readdir(directoryPath, async (err, fileNames): Promise<void> => {
      if (err) {
        reject(err);
        return;
      }
      if (fileNames === undefined) {
        return;
      }

      const removeFile = (fileName: string): Promise<Stats> => {
        const fullPath = path.join(directoryPath, fileName);
        return new Promise((resolve, reject) => {
          fileSystem.lstat(fullPath, async (err, stats) => {
            if (err) {
              reject(err);
              return;
            }
            if (stats === undefined) {
              return;
            }

            if (stats.isFile()) {
              await new Promise((resolve, reject) => {
                fileSystem.unlink(fullPath, err => {
                  if (err) {
                    reject(err);
                    return;
                  }
                  resolve(true);
                });
              });
            } else if (stats.isDirectory()) {
              await rmdirRecursively(fileSystem, fullPath);
            }

            resolve(stats);
          });
        });
      };

      // Remove each file/directory in parallel.
      Promise.all(fileNames.map(removeFile))
        .then(() => resolve())
        .catch(err => reject(err));
    });
  });
};

/**
 * Removes a directory recursively.
 *
 * BrowserFS's `rmdir` function is unable to remove non-empty directories despite being modelled
 * after Node.js's file system API. In Node.js, the `rmdir` function takes in an `options` object
 * where recursive directory removal can be set (https://nodejs.org/api/fs.html#fspromisesrmdirpath-options).
 * The BrowserFS equivalent however does not support these options and will fail with ENOTEMPTY
 * when trying to remove a non-empty directory.
 *
 * @param fileSystem    The file system instance.
 * @param directoryPath The path of the directory to be removed.
 */
export const rmdirRecursively = (fileSystem: FSModule, directoryPath: string): Promise<void> => {
  return new Promise((resolve, reject) => {
    rmFilesInDirRecursively(fileSystem, directoryPath)
      .then(() => {
        fileSystem.rmdir(directoryPath, err => {
          if (err) {
            reject(err);
            return;
          }
          resolve();
        });
      })
      .catch(err => reject(err));
  });
};

/**
 * Writes a file recursively by creating all parent directories if they do not exist.
 *
 * BrowserFS's `mkdir` function is unable to recursively create directories despite being modelled
 * after Node.js's file system API. In Node.js, the `mkdir` function takes in an `options` object
 * where recursive directory creation can be set (https://nodejs.org/api/fs.html#fspromisesmkdirpath-options).
 * The BrowserFS equivalent however does not support these options and will fail with ENOENT
 * when trying to create directories recursively.
 *
 * @param fileSystem   The file system instance.
 * @param filePath     The path of the file to be written to.
 * @param fileContents The contents of the file to be written to.
 */
export const writeFileRecursively = (
  fileSystem: FSModule,
  filePath: string,
  fileContents: string
): Promise<void> => {
  return new Promise((resolve, reject) => {
    // Create directories along the path if they do not exist.
    // Remove empty path segments.
    const pathSegments = filePath.split(path.sep).filter(path => path !== '');
    // The last path segment is not a directory but the name of the file, so we ignore it.
    for (let i = 1; i < pathSegments.length; i++) {
      const dirPath = '/' + path.join(...pathSegments.slice(0, i));
      const promise: Promise<void> = new Promise((resolve, reject) => {
        // Check whether the directory path already exists to prevent overwriting of existing files & directories.
        fileSystem.exists(dirPath, dirPathExists => {
          if (dirPathExists) {
            resolve();
            return;
          }

          fileSystem.mkdir(dirPath, 777, err => {
            if (err) {
              reject(err);
              return;
            }

            resolve();
          });
        });
      });
      promise.catch(err => reject(err));
    }

    // Write to the file.
    fileSystem.writeFile(filePath, fileContents, err => {
      if (err) {
        reject(err);
        return;
      }

      resolve();
    });
  });
};
