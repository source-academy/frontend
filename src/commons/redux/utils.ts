import {
  ActionCreatorWithOptionalPayload,
  ActionCreatorWithoutPayload,
  ActionCreatorWithPreparedPayload,
  createAction
} from '@reduxjs/toolkit';
import * as Sentry from '@sentry/browser';
import { SagaIterator } from 'redux-saga';
import { StrictEffect, takeEvery } from 'redux-saga/effects';

/**
 * Creates actions, given a base name and base actions
 * @param baseName The base name of the actions
 * @param baseActions The base actions. Use a falsy value to create an action without a payload.
 * @returns An object containing the actions
 */
export function createActions<BaseName extends string, BaseActions extends Record<string, any>>(
  baseName: BaseName,
  baseActions: BaseActions
) {
  return Object.entries(baseActions).reduce(
    (res, [name, func]) => ({
      ...res,
      [name]: func
        ? createAction(`${baseName}/${name}`, (...args: any) => ({ payload: func(...args) }))
        : createAction(`${baseName}/${name}`)
    }),
    {} as {
      [K in keyof BaseActions]: K extends string
        ? BaseActions[K] extends (...args: any) => any
          ? ActionCreatorWithPreparedPayload<
              Parameters<BaseActions[K]>,
              ReturnType<BaseActions[K]>,
              `${BaseName}/${K}`
            >
          : ActionCreatorWithoutPayload<`${BaseName}/${K}`>
        : never;
    }
  );
}

export function combineSagaHandlers<
  TActions extends Record<
    string,
    ActionCreatorWithPreparedPayload<any, any> | ActionCreatorWithoutPayload<any>
  >
>(
  actions: TActions,
  handlers: {
    // TODO: Maybe this can be stricter? And remove the optional type after
    // migration is fully done
    [K in keyof TActions]?: (action: ReturnType<TActions[K]>) => SagaIterator;
  },
  others?: (takeEvery: typeof saferTakeEvery) => SagaIterator
): () => SagaIterator {
  const sagaHandlers = Object.entries(handlers).map(([actionName, saga]) =>
    saferTakeEvery(actions[actionName], saga)
  );
  return function* (): SagaIterator {
    yield* sagaHandlers;
    if (others) {
      const obj = others(saferTakeEvery);
      while (true) {
        const { done, value } = obj.next();
        if (done) break;
        yield value;
      }
    }
  };
}

export function saferTakeEvery<
  Action extends
    | ActionCreatorWithOptionalPayload<any>
    | ActionCreatorWithPreparedPayload<any[], any>
>(actionPattern: Action, fn: (action: ReturnType<Action>) => Generator<StrictEffect<any>>) {
  function* wrapper(action: ReturnType<Action>) {
    try {
      yield* fn(action);
    } catch (error) {
      handleUncaughtError(error);
    }
  }

  return takeEvery(actionPattern.type, wrapper);
}

function handleUncaughtError(error: any) {
  if (process.env.NODE_ENV === 'development') {
    // react-error-overlay is a "special" package that's automatically included
    // in development mode by CRA

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    import('react-error-overlay').then(reo => reo.reportRuntimeError(error));
  }
  Sentry.captureException(error);
  console.error(error);
}
