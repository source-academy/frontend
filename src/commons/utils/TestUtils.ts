import React from 'react';
import renderer from 'react-test-renderer';
import { createRenderer } from 'react-test-renderer/shallow';

export const shallowRender = (element: React.ReactElement) => {
  const renderer = createRenderer();
  renderer.render(element);
  return renderer.getRenderOutput();
};

export const renderTreeJson = (element: React.ReactElement) => {
  return renderer.create(element).toJSON();
};

export const renderTree = (element: React.ReactElement) => {
  return renderer.create(element);
};

/**
 * Recursively traverses a nested object and returns all matches.
 *
 * Used in traversing nested objects such as `shallowRender`'s output.
 */
export function deepFilter<T>(
  nestedObject: T,
  matchFn: (e: T) => boolean,
  getChildren: (e: T) => T[]
) {
  const matches: any[] = [];

  function helper(obj: T) {
    if (matchFn(obj)) {
      matches.push(obj);
    }

    const children = getChildren(obj);
    if (children && Array.isArray(children)) {
      children.forEach(e => {
        if (e) {
          helper(e);
        }
      });
    }
  }

  helper(nestedObject);
  return matches;
}
