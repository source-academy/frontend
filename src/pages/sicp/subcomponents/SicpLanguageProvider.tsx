import { createContext, useCallback, useContext, useState } from 'react';
import {
  getSicpLanguageFromLocalStorage,
  SICP_DEFAULT_LANGUAGE,
  type SicpSupportedLanguage
} from 'src/features/sicp/utils/SicpUtils';

type SicpLanguageContext = {
  sicpLanguage: SicpSupportedLanguage;
  setSicpLanguage: (lang: SicpSupportedLanguage) => void;
};

const sicpLanguageContext = createContext<SicpLanguageContext | undefined>(undefined);

export const useSicpLanguageContext = (): SicpLanguageContext => {
  const context = useContext(sicpLanguageContext);
  if (!context) {
    throw new Error('useSicpLanguageContext must be used inside an SicpLanguageContextProvider');
  }

  return context;
};

export const SicpLanguageContextProvider: React.FC<{ children: React.ReactNode }> = ({
  children
}) => {
  const [lang, setLang] = useState<SicpSupportedLanguage>(
    getSicpLanguageFromLocalStorage() ?? SICP_DEFAULT_LANGUAGE
  );

  const handleLangChange = useCallback((newLang: SicpSupportedLanguage) => {
    setLang(newLang);
  }, []);

  return (
    <sicpLanguageContext.Provider
      value={{
        sicpLanguage: lang,
        setSicpLanguage: handleLangChange
      }}
    >
      {children}
    </sicpLanguageContext.Provider>
  );
};
