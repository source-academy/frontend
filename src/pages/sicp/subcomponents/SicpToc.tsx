import { Tree, TreeNodeInfo } from '@blueprintjs/core';
import { cloneDeep } from 'lodash';
import * as React from 'react';
import { useState } from 'react';
import { useHistory } from 'react-router';

import toc from '../../../features/sicp/data/toc.json';

type TocProps = OwnProps;

type OwnProps = {
  handleCloseToc?: () => void;
};

/**
 * Table of contents of SICP.
 */
const SicpToc: React.FC<TocProps> = props => {
  const [sidebarContent, setSidebarContent] = useState(toc as TreeNodeInfo[]);
  const history = useHistory();

  const handleNodeExpand = (_node: TreeNodeInfo, path: integer[]) => {
    const newState = cloneDeep(sidebarContent);
    Tree.nodeFromPath(path, newState).isExpanded = true;
    setSidebarContent(newState);
  };

  const handleNodeCollapse = (_node: TreeNodeInfo, path: integer[]) => {
    const newState = cloneDeep(sidebarContent);
    Tree.nodeFromPath(path, newState).isExpanded = false;
    setSidebarContent(newState);
  };

  const handleNodeClicked = React.useCallback(
    (node: TreeNodeInfo) => {
      if (props.handleCloseToc) {
        props.handleCloseToc();
      }
      history.push('/sicpjs/' + String(node.nodeData));
    },
    [history, props]
  );

  return (
    <div className="sicp-toc">
      <Tree
        className="sicp-toc-tree"
        contents={sidebarContent}
        onNodeClick={handleNodeClicked}
        onNodeCollapse={handleNodeCollapse}
        onNodeExpand={handleNodeExpand}
      />
    </div>
  );
};

export default SicpToc;
