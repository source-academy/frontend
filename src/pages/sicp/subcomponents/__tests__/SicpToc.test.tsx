import { MemoryRouter } from 'react-router';
import { renderTreeJson } from 'src/commons/utils/TestUtils';

import { SicpLanguageContextProvider } from '../SicpLanguageProvider';
import SicpToc from '../SicpToc';

test('Sicp toc renders correctly', async () => {
  const props = {
    handleCloseToc: () => {}
  };

  const tree = await renderTreeJson(
    <MemoryRouter>
      <SicpLanguageContextProvider>
        <SicpToc {...props} />
      </SicpLanguageContextProvider>
    </MemoryRouter>
  );
  expect(tree).toMatchSnapshot();
});
