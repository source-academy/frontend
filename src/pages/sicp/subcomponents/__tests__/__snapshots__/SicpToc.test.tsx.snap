// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sicp toc renders correctly 1`] = `
<div
  className="sicp-toc"
>
  <div
    className="sicp-content"
  >
    <div
      className="bp5-non-ideal-state bp5-non-ideal-state-vertical"
    >
      <div
        className="bp5-non-ideal-state-visual"
        style={
          {
            "fontSize": "48px",
            "lineHeight": "48px",
          }
        }
      >
        <div
          aria-label="loading"
          aria-valuemax={100}
          aria-valuemin={0}
          className="bp5-spinner"
          role="progressbar"
        >
          <div
            className="bp5-spinner-animation"
          >
            <svg
              height={50}
              strokeWidth="8.00"
              viewBox="1.00 1.00 98.00 98.00"
              width={50}
            >
              <path
                className="bp5-spinner-track"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
              />
              <path
                className="bp5-spinner-head"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                pathLength={280}
                strokeDasharray="280 280"
                strokeDashoffset={210}
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        className="bp5-non-ideal-state-text"
      >
        <h4
          className="bp5-heading"
        >
          Loading Content
        </h4>
      </div>
    </div>
  </div>
</div>
`;
